tunnistamo/tests/test_oidc_provider_backchannel_logout.py

```diff
@@ -0,0 +1,203 @@
+
+    assert logout_token_decoded['aud'] == client.client_id
+    assert logout_token_decoded['sub'] == str(user.uuid)
+    assert logout_token_decoded['events'] == {
+        'http://schemas.openid.net/event/backchannel-logout': {}
+    }
+    if sid:
+        assert logout_token_decoded['sid'] == sid
+
+
+@pytest.mark.django_db
+def test_openid_configuration_announces_backchannel_logout_support():
+    client = DjangoTestClient()
+    response = client.get('/.well-known/openid-configuration/')
+    response_data = response.json()
+
+    assert response.status_code == 200
+    assert response_data['backchannel_logout_supported'] is True
+    assert response_data['backchannel_logout_session_supported'] is True
+
+
+@pytest.mark.django_db
+def test_create_logout_token(rsa_key, oidcclient_factory):
+    oidc_client = oidcclient_factory(redirect_uris=['https://example.com'])
+
+    logout_token = create_logout_token(oidc_client, iss='test-iss', sub='test-sub', sid='test-sid')
+    logout_token_decoded = jwt.decode(logout_token, verify=False)
+
+    assert logout_token_decoded['iss'] == 'test-iss'
+    assert logout_token_decoded['aud'] == oidc_client.client_id
+    assert logout_token_decoded['sub'] == 'test-sub'
+    assert logout_token_decoded['events'] == {
+        'http://schemas.openid.net/event/backchannel-logout': {}
+    }
+    assert logout_token_decoded['sid'] == 'test-sid'
+
+
+@pytest.mark.django_db
+@httprettified
+def test_send_backchannel_logout_to_apis_no_apis_in_scope(user, oidcclient_factory):
+    oidc_client = oidcclient_factory(redirect_uris=['https://example.com'])
+    token = create_token(user, oidc_client, [])
+
+    send_backchannel_logout_to_apis_in_token_scope(token, request=RequestFactory().get('/'))
+
+    assert len(httpretty.latest_requests) == 0
+
+
+@pytest.mark.django_db
+@httprettified
+def test_send_backchannel_logout_to_apis_one_api_in_scope_no_url_in_api(user):
+    oidc_client = create_oidc_clients_and_api()
+    api_scope = ApiScope.objects.filter(allowed_apps=oidc_client).first()
+    token = create_token(user, oidc_client, [api_scope.identifier])
+
+    api_scope.api.backchannel_logout_url = None
+    api_scope.api.save()
+
+    send_backchannel_logout_to_apis_in_token_scope(token, request=RequestFactory().get('/'))
+
+    assert len(httpretty.latest_requests) == 0
+
+
+@pytest.mark.django_db
+def test_send_backchannel_logout_to_apis_one_api_in_scope_ignore_http_error(user):
+    oidc_client = create_oidc_clients_and_api()
+    api_scope = ApiScope.objects.filter(allowed_apps=oidc_client).first()
+    token = create_token(user, oidc_client, [api_scope.identifier])
+
+    api_scope.api.backchannel_logout_url = 'invalid url'
+    api_scope.api.save()
+
+    send_backchannel_logout_to_apis_in_token_scope(token, request=RequestFactory().get('/'))
+
+
+@pytest.mark.django_db
+@httprettified
+def test_send_backchannel_logout_to_apis_one_api_in_scope(user):
```

Comment:
To me it seems this test doesn't cover anything more than the next test (`test_send_backchannel_logout_to_apis_two_apis_in_scope`), so perhaps this test could be removed?

Reply:
Maybe it's a philosophic question, but if we only test with two APIs, do we know that it works with only one API? 😄

Reply:
It could very well be a valid question. If you think it's beneficial to have both of these tests, then it's ok to keep them.